### DISCUSSION TITLE
単体テストの考え方/使い方 2.2〜2.3

### DISCUSSION THREAD

**Discussion by dameleon** [atman-inc/training#92] [General]
- 2.2 古典学派及びロンドン学派が考える単体テスト
- 2.3 単体テストにおける古典学派とロンドン学派の違い

上記2章分のディスカッション。
不明点、相談点、確認点、勉強になったポイントなどは各自ディスカッションのコメントをお願いします。

進行: @dameleon

**Comment by koushisa** [comment]
## 2.2

### 可変依存, 不変依存

- 説明が遠回しだと感じた
- 値オブジェクト (value object)
  - いろいろ方言があるが(PofEAA, DDD)本書では"単なるimmutableオブジェクト"として扱っているっぽい
  - 初見の人は理解がむずかしそう
  - いまいち納得感がないというか著者の定義と自分の理解が同じなのかあやふやなまま過ぎ去った
    - きっと`Product.Shampoo`にsetterや依存関係がない = 不変であるということを言いたかったのだとおもう

##  2.3

- P47のtipsは大事だとおもった
  - 単体テストの粒度の話
    - 「単体テストは1単位のコードを検証するのではなく1単位の振る舞いを検証する」
    - 「ビジネスサイドの人たちが有用とであると考えるなにか」
  - これをフロントエンドに当てはめると自然と結合テスト的な考え方に行き着く
    - 最近だとtesting trophyと言われてるやつ
  - もちろん共通関数や取り回される関数は単体テストはあったほうがよいが、フロントエンドの振る舞いはやっぱり結合観点になる
  - 自分がフロントエンドのテストを考えるときは
    - ユーザーのインタラクション単位(=Action単位)がよいとおもっている
      - ReadよりもWriteにビジネスロジックが集まりがちなため
    - 状態遷移(Reducer)の入力やコンポーネントのpropsはモックしない
      - enzymeよりもtesting libraryを好む
	  - Action→状態遷移→UIの再レンダリングを1シナリオとして捉える
	    - hooksや関数分割を頑張ってテストを書くよりもコスパがいい
	    - Actionに対するUI(コンポーネント)の振る舞いが担保されていれば、コンポーネントが内部的に利用しているhooksの動作もある程度は担保される
  - フロントエンド文脈で同じ本を読んで同じような感想を持った人がいた
    - https://zenn.dev/akfm/articles/frontend-unit-testing
- 自分は(フロントエンドにおいては)テスト駆動よりも、テストラスト派
  - https://blog.jnito.com/entry/2023/02/16/171810
  - > 雑に動くものを作る↓
  - > 見た目をきれいにする＆機能を作り込む↓
  - > テストを書く↓
  - > リファクタリングする

**Comment by xxManatoxx** [comment]
- 勉強になったポイント
  - ロンドン学派のメリット
    - より細かな粒度で検証できる
    - 依存関係が複雑でも簡単にテストできる
    - テストが失敗した際にどこに問題があったのか正確に見つけられる
  - 上記メリットを上回る懸念を感じました。
    - 依存関係のクラスや機能に変更があった場合、漏れなくモックを修正できるか
      - 依存関係に変更があった場合にMock化している箇所を全て修正する必要があり、Mockを直さなければテストで検知できない状態になっていると漏れがありそうだなと感じました
      - エディタの検索機能などで探せば漏れは減らせるとも思いますが、テストで検知できなくなるのは恐怖を感じます
    - 上記の懸念点が他2つのメリットを大きく上回っていると感じたので個人的にあまりMock化し過ぎるのはやりたくないなと思ってしまいました。

**Comment by nmurakami0** [comment]
2.2 古典学派及びロンドン学派が考える単体テスト

メモ

- 共有依存
    - テストケース間で共有される依存
- ロンドン学派でもimmutableならテストダブルに置き換える必要はない
    - immutableなオブジェクト: 値オブジェクト
- 協力者オブジェクト(collaborator)は共有依存か可変依存のどちらか
    - 古典学派なら共有依存でなければ良い
- シングルトンは共有依存だがプロセス外依存ではない
    - (本番環境では共通のオブジェクトにしていても、テストケースごとに新しいインスタンスを作成すれば共有依存にならなくプライベート依存になる)
- プロセス外依存だからといって共有依存とは限らない
    - ただ、プロセス外依存だが共有依存でないものは実際のプロジェクトではほとんどない

感想

- rspecだとdatabase rewindで共有依存を無くしている
- spec書いていてcollaboratorの準備がかなり大変
    - factoryがややこしくなる
    - 振る舞いがややこしいから仕方がない？

2.3 単体テストにおける古典学派とロンドン学派の違い

メモ

- 単体テストでは、a unit of codeを検証するのではなくa unit of behaviorを検証する
    - よって、コードの粒度を細かくすることは単体テストにおいてはそこまで意味がない
    - 振る舞いが何かを伝えられるのが重要。cohesionを高めると非開発者にも伝えやすくなる
- ロンドン学派のようにcollaboratorをモックにすると複雑な依存関係を持つテスト対象をテストしやすくなる
    - ただ、そもそも、複雑な依存関係を構築しなくする方が重要
- 古典学派スタイルだと依存している対象にバグがあってもテストが失敗する
    - しかし、単体テストを頻繁に実施していればどこでバグがあるかはすぐわかる
    - また、多くのテストに影響を与えるcollaboratorは重要な価値があると言える
        - e.g., よく使われるサービスクラス
    - 設計や変更などでコードを扱う際に把握していると有用な情報になる
- TDD
    - ロンドン学派によるテストスタイルは、外側から内側に向かう
    - (粒度が大きい振る舞いを表すクラスをモックにすれば良いから？)
    - 何で実装の詳細に結びつく？ -> クラス単位でテストするから?

感想: 古典学派はリアリスティック。原理主義ではなく、結果として有益につながることをプラスに捉えている

**Comment by nmurakami0** [comment]
value object: The concept of identifier equality refers to entities, whereas the concept of structural equality - to value objects. 

https://enterprisecraftsmanship.com/posts/entity-vs-value-object-the-ultimate-list-of-differences/

**Comment by yutafukuchi-1015** [comment]
#### 感想
- 関数(クラス)内の責務をなるべく1つ/減らした方がテスト/確認しやすい
  - 仕様と実装の差分をテストしやすい
  - テストエラー原因がどこなのか明確になる
  - 仕様の理解のためにも読みやすい(金田さんが先日仰っていたことだが)
  - 既存の実装への影響が出た場合検知できる
- FEに効率的にテスト導入するには、既存実装への影響が出ないようにロジックにテストを書くのが良さそう
